mkdocs/docs/bindings/c-api.md

```diff
@@ -1,9 +1,224 @@
 # C API bindings
 
-<!-- TODO(??): overview, advantages/disadvantages to using raw C APIs -->
+IREE provides a low level C API for its runtime[^1], which can be used directly or
+through higher level APIs and language bindings built on top of it.
+
+API header files are organized by runtime component:
+
+| Component header file                                                       | Overview                                                                  |
+|-----------------------------------------------------------------------------|---------------------------------------------------------------------------|
+| [iree/base/api.h](https://github.com/google/iree/blob/main/iree/base/api.h) | Core API, type definitions, ownership policies, utilities                 |
+| [iree/vm/api.h](https://github.com/google/iree/blob/main/iree/vm/api.h)     | VM APIs: loading modules, I/O, calling functions                          |
+| [iree/hal/api.h](https://github.com/google/iree/blob/main/iree/hal/api.h)   | HAL APIs: device management, synchronization, accessing hardware features |
+
+The [iree/samples/](https://github.com/google/iree/tree/main/iree/samples)
+directory demonstrates several ways to use IREE's C API.
 
 ## Prerequisites
 
-<!-- TODO(??): build from source, vcpkg -->
+To use IREE's C API, you will need to build the runtime
+[from source](../building-from-source/getting-started.md). The
+[iree-template-cpp](https://github.com/iml130/iree-template-cpp) community
+project also shows how to integrate IREE into an external project using
+CMake[^2].
+
+## Concepts
+
+By default, IREE uses a Virtual Machine (VM) at runtime to interpret program
+instructions on the host system. VM instructions may also be lowered further to
+LLVM IR, C, or other representations for static or resource constrained
+deployment.
+
+The VM supports generic operations like loads, stores, arithmetic, function
+calls, and control flow. It builds streams of more complex program logic and
+dense math into command buffers that are dispatched to hardware backends
+through the Hardware Abstraction Layer (HAL) interface.
+
+Most interaction with IREE's C API involves either the VM or the HAL.
+
+<!-- TODO(scotttodd): diagrams -->
+
+### IREE VM
+
+* VM _instances_ can serve multiple isolated execution _contexts_
+* VM _contexts_ are effectively sandboxes for loading modules and running
+  programs
+* VM _modules_ provide extra functionality to execution _contexts_, such as
+  access to hardware accelerators through the HAL
+
+### IREE HAL
+
+* HAL _drivers_ are used to enumerate and create HAL _devices_
+* HAL _devices_ interface with hardware, such as by allocating device memory,
+  preparing executables, recording and dispatching command buffers, and
+  synchronizing with the host
```

Comment:
worth calling out something like `*  HAL _buffers_ and _buffer views_ represent storage and shaped/typed views into that storage (aka "tensors")` (as that's what most people will be messing with as I/O)